### PR TITLE
[feature][console] Add toggle comment action in the python console

### PR DESCRIPTION
**Pull request by YoannQDQ**
## Description

Replace the comment/uncomment actions of the embedded Python Editor with a toggle comment action, whose beahavior is similar to VSCode toggle comment.

Shortcut for this action is set to `Ctrl+:`

toolbar

When toggling comment on a single line, the behavior is as follow :
- If the line starts with  `"#"` , or ` "# "` (space matters), remove it
- Else add  `"# "` before the line
- The text selection is kept as is, so triggering the action twice reverts to the original state

When toggling comment on multiple lines, the behavior is as follow :
- If at least one **non empty** line does not start with a `"#"`, add `"# "` at the start of each **non empty** line
- Otherwise, remove the first `"#" `or` "# "` of every non empty line

toggle

Also add the toggle comment action to the **Processing Script Editor**.

Toggle comment via the `Ctrl+:` shortcut is also available in **Project Macro** and **Expression Editors** (without button though)

scriptedit

### CODE REVIEW

File: python/plugins/processing/script/ScriptEdit.py [hunk 1]
```diff
@@ -55,3 +55,53 @@ def initShortcuts(self):
                                                            + Qt.Key_Space), self)
         self.shortcutAutocomplete.setContext(Qt.WidgetShortcut)
         self.shortcutAutocomplete.activated.connect(self.autoComplete)
+
+    def toggleComment(self):
```
Comment (nyalldawson) [inline python/plugins/processing/script/ScriptEdit.py:59]: How do you feel about moving this logic to sit in the base QgsCodeEditorPython class instead? Then we could utilise it from everywhere without the duplicate code (even in non-python code parts where QgsCodeEditorPython is used, e.g. in the project macros configuration)

**Comment by nyalldawson** [comment]
Great contribution @YoannQDQ ! This is a very nice usability tweak to see

File: python/plugins/processing/script/ScriptEdit.py (hunk 1 above)
Comment (YoannQDQ) [reply python/plugins/processing/script/ScriptEdit.py:59]: That would indeed be better. I will update this PR.

**Comment by YoannQDQ** [comment]
Refactored to embed the toggleComment code in QgsCodeEditorPython.

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 2]
```diff
@@ -185,6 +186,19 @@ void QgsCodeEditorPython::initializeLexer()
   runPostLexerConfigurationTasks();
 }
 
+void QgsCodeEditorPython::keyPressEvent(QKeyEvent *event)
+{
+  bool ctrl = event->modifiers() & Qt::ControlModifier;
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:191]:
```suggestion
  const bool ctrlModifier = event->modifiers() & Qt::ControlModifier;
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 3]
```diff
@@ -185,6 +186,19 @@ void QgsCodeEditorPython::initializeLexer()
...
+
+  if(ctrl && event->key() == Qt::Key_Colon)
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:193]:
```suggestion
  if ( ctrlModifier && event->key() == Qt::Key_Colon )
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 4]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+  if ( isReadOnly() )
+  {
+    return;
+  }
+
+  beginUndoAction();
+  int startLine, startPos, endLine, endPos;
+  if ( hasSelectedText() )
+    getSelection(&startLine, &startPos, &endLine, &endPos);
+  else
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:270]:
```suggestion
  if ( hasSelectedText() )
  {
    getSelection(&startLine, &startPos, &endLine, &endPos);
  }
  else
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 5]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+    getCursorPosition(&startLine, &startPos);
+    endLine = startLine;
+    endPos = startPos;
+  }
+
+  // Check comment state and minimum indentation for each selected line  
+  bool allEmpty = true;
+  bool allCommented = true;
+  int minIndentation = -1;
+  for ( int line=startLine; line<=endLine; line++ )
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:281]:
```suggestion
  for ( int line = startLine; line <= endLine; line++ )
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 6]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+  {
+    auto stripped = text(line).trimmed();
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:283]:
```suggestion
    const QString stripped = text( line ).trimmed();
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 7]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+    if ( !stripped.isEmpty() )
+    {
+      allEmpty = false;
+      if ( !stripped.startsWith("#") )
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:287]:
```suggestion
      if ( !stripped.startsWith( '#' ) )
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 8]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+      {
+        allCommented=false;
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:289]:
```suggestion
        allCommented = false;
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 9]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+      }
+      if ( minIndentation == -1 || minIndentation > indentation(line) )
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:291]:
```suggestion
      if ( minIndentation == -1 || minIndentation > indentation( line ) )
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 10]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+      {
+        minIndentation = indentation(line);
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:293]:
```suggestion
        minIndentation = indentation( line );
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 11]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+  // Special case, only empty lines
+  if ( allEmpty )
+  {
+    return;
+  }
+
+  // Selection shift to keep the same selected text after a # is added/removed
+  int delta = 0;
+
+  for ( int line=startLine; line<=endLine; line++ )
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:307]:
```suggestion
  for ( int line = startLine; line <= endLine; line++ )
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 12]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+  {
+    auto stripped = text(line).trimmed();
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:309]:
```suggestion
    const QString stripped = text( line ).trimmed();
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 13]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+
+    // Empty line
+    if ( stripped.isEmpty() )
+    {
+      continue;
+    }
+
+    if ( !allCommented )
+    {
+      insertAt("# ", line, minIndentation);
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:319]:
```suggestion
      insertAt( QStringLiteral( "# " ), line, minIndentation );
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 14]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+      delta = -2;
+    }
+    else
+    {
+      if ( !stripped.startsWith("#") )
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:324]:
```suggestion
      if ( !stripped.startsWith( '#' ) )
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 15]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+      {
+        continue;
+      }
+      if ( stripped.startsWith("# ") )
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:328]:
```suggestion
      if ( stripped.startsWith( QLatin1String( "# " ) ) )
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 16]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+      {
+        delta = 2;
+      }
+      else
+      {
+        delta = 1;
+      }
+      setSelection(line, indentation(line), line, indentation(line) + delta);
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:336]:
```suggestion
      setSelection( line, indentation( line ), line, indentation( line ) + delta );
```

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 17]
```diff
@@ -242,6 +256,92 @@ void QgsCodeEditorPython::searchSelectedTextInPyQGISDocs()
...
+      removeSelectedText();
+    }
+  }
+
+  endUndoAction();
+  setSelection(startLine, startPos - delta, endLine, endPos - delta);
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:342]:
```suggestion
  setSelection( startLine, startPos - delta, endLine, endPos - delta );
```

File: src/gui/codeeditors/qgscodeeditorpython.h [hunk 18]
```diff
@@ -84,10 +84,24 @@ class GUI_EXPORT QgsCodeEditorPython : public QgsCodeEditor
...
+
   protected:
 
     void initializeLexer() override;
 
+    /**
+     * Override QgsCodeEdito::keyPressEvent to provide custom shortcuts
+     *
+     * \since QGIS 3.30
+     */
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.h:102]:
```suggestion
```

No doxygen is necessary for overridden methods

File: src/gui/codeeditors/qgscodeeditorpython.cpp [hunk 19]
```diff
@@ -256,9 +356,9 @@ const char *QgsQsciLexerPython::keywords( int set ) const
 {
   if ( set == 1 )
   {
-    return "True False and as assert break class continue def del elif else except exec "
+    return "True False and as assert break class continue def del elif else except "
            "finally for from global if import in is lambda None not or pass "
-           "print raise return try while with yield";
+           "raise return try while with yield async await nonlocal";
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.cpp:361]: why these changes?

Comment (YoannQDQ) [reply src/gui/codeeditors/qgscodeeditorpython.cpp:361]: print and exec are no longer keywords for Python2. However, async, await and nonlocal were added as keywords in Python3

File: src/gui/codeeditors/qgscodeeditorpython.h (hunk 18 above)
Comment (YoannQDQ) [reply src/gui/codeeditors/qgscodeeditorpython.h:102]: Duly noted

**Comment by YoannQDQ** [comment]
@nyalldawson: Thank you for your review. I noticed most of the fixes you suggested were style errors. Could you tell me which style guide/beautifier/formatter is used in QGIS?

File: src/gui/codeeditors/qgscodeeditorpython.cpp (hunk 19 above)
Comment (nyalldawson) [reply src/gui/codeeditors/qgscodeeditorpython.cpp:363]: Ok, makes sense 👍

**Comment by nyalldawson** [comment]
@YoannQDQ the scripts/prepare_commit.sh script is designed to be used as a git pre commit hook and automatically applies all formatting tools needed to keep QGIS ci happy.

Ideally this should be set as your local git commit hook, but you can also run it manually if you prefer.

There's also scripts/astyle_all.sh and scripts/sipify_all.sh which will bulk update code formatting (it's very slow, as it runs on all files in the repo, but is useful for fixing formatting added before the pre commit script was setup).

**Comment by nyalldawson** [comment]
@YoannQDQ this all looks good to me -- just need to get the formatting tests passing and I'll merge! Thanks!

**Comment by YoannQDQ** [comment]
> @YoannQDQ this all looks good to me -- just need to get the formatting tests passing and I'll merge! Thanks!

My mistake, I had manually edited the sip file (despite the "auto_generated" in the path which was a dead giveaway...).
I set up the pre-commit hook (wish I had done so sooner) and manually ran astyle_all.sh and sipify_all.sh. Hopefully all tests should pass now.

**Comment by YoannQDQ** [comment]
🥳 All checks have passed!  Would it now be possible to approve this PR for merging? Thanks

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Thanks @YoannQDQ !
